filament/backend/src/opengl/OpenGLProgram.cpp

```diff
@@ -35,43 +35,92 @@ using namespace backend;
 
 static void logCompilationError(utils::io::ostream& out,
         backend::Program::Shader shaderType, const char* name,
-        GLuint shaderId, std::string_view source) noexcept;
+        GLuint shaderId) noexcept;
 
 static void logProgramLinkError(utils::io::ostream& out,
         const char* name, GLuint program) noexcept;
 
-OpenGLProgram::OpenGLProgram(OpenGLDriver* gl, const Program& programBuilder) noexcept
-        :  HwProgram(programBuilder.getName()), mIsValid(false) {
+OpenGLProgram::OpenGLProgram(OpenGLContext& context, Program&& programBuilder) noexcept
+        :  HwProgram(programBuilder.getName()) {
 
-    using Shader = Program::Shader;
+    auto samplerGroupInfo = std::move(programBuilder.getSamplerGroupInfo());
+    auto uniformBlockInfo = std::move(programBuilder.getUniformBlockInfo());
 
-    const auto& shadersSource = programBuilder.getShadersSource();
-    OpenGLContext& context = gl->getContext();
+    // this cannot fail because we check compilation status after linking the program
+    // shaders[] is filled with id of shader stages present.
+    OpenGLProgram::compileShaders(context, programBuilder.getShadersSource(), gl.shaders);
+
+    // link the program, this also cannot fail because status is checked later.
+    // TODO: defer this until beginRenderPass()
+    gl.program = OpenGLProgram::linkProgram(gl.shaders);
+
+    // check status of program linking and shader compilation, logs error and free all resources
+    // in case of error.
+    // TODO: defer this until first use
+    bool success = OpenGLProgram::checkProgramStatus(name.c_str_safe(),
+            gl.program, gl.shaders);
+
+    // Failing to compile a program can't be fatal, because this will happen a lot in
+    // the material tools. We need to have a better way to handle these errors and
+    // return to the editor. Also note the early "return" statements in this function.
```

Comment:
Hm, I don't see any `return` statements in `OpenGLProgram::OpenGLProgram`

Reply:
good catch.